GH-38837: [Format] Add the specification for statistics schema (#45058)
apache/arrow: docs/source/format/StatisticsSchema.rst

  array.

Schema
======

This specification provides only the schema for statistics. This is
the canonical schema to represent statistics about an Apache Arrow
dataset as Apache Arrow data.

Here is the outline of the schema for statistics::

    struct<
      column: int32,
      statistics: map<
        key: dictionary<
          indices: int32,
          dictionary: utf8
        >,
        items: dense_union<...all needed types...>
      >
    >

Here is the details of top-level ``struct``:

.. list-table::
   :header-rows: 1

   * - Name
     - Data type
     - Nullable
     - Notes
   * - ``column``
     - ``int32``
     - ``true``
     - The zero-based column index, or null if the statistics
       describe the whole table or record batch.

       The column index is computed as the same rule used by
       :ref:`ipc-recordbatch-message`.
   * - ``statistics``
     - ``map``
     - ``false``
     - Statistics for the target column, table or record batch. See
       the separate table below for details.

Here is the details of the ``map`` of the ``statistics``:

.. list-table::
   :header-rows: 1

   * - Key or items
     - Data type
     - Nullable
     - Notes
   * - key

>      - ``dictionary<indices: int32, dictionary: utf8>``

Author: Ah, I didn't notice the point. Thanks.

The current proposed specification doesn't focus on any transports/protocols/APIs/.... So this may not be a problem. If this representation doesn't match for our IPC formats, users just don't use this.

     - ``false``
     - Statistics key is string. Dictionary is used for
       efficiency. Different keys are assigned for exact value and
       approximate value. Also see the separate description below for
       statistics key.
   * - items
     - ``dense_union``
     - ``false``
     - Statistics value is dense union. It has at least all needed
       types based on statistics kinds in the keys. For example, you
       need at least ``int64`` and ``float64`` types when you have a
       ``int64`` distinct count statistic and a ``float64`` average
       byte width statistic. Also see the separate description below
       for statistics key.

       We don't standardize field names for the dense union because we
       can access to proper field by type code not name. So we can use
       any valid name for fields.

>        TODO: Should we standardize field names?

Author: Related discussion: https://github.com/apache/arrow/pull/43553/files#r1871755575

>        TODO: Should we standardize field names?

Author: There is no objection for not standardizing field names.

So I don't standardize field names.

:ref:`format_metadata`. It can be used multiple times in a key.

The ``ARROW`` pattern is a reserved namespace for pre-defined
statistics keys. User-defined statistics must not use it.
For example, you can use your product name as namespace
such as ``MY_PRODUCT:my_statistics:exact``.

Here are pre-defined statistics keys:

.. list-table::
   :header-rows: 1

   * - Key
     - Data type
     - Notes
   * - ``ARROW:average_byte_width:exact``
     - ``float64``
     - The average size in bytes of a row in the target
       column. (exact)
   * - ``ARROW:average_byte_width:approximate``
     - ``float64``
     - The average size in bytes of a row in the target
       column. (approximate)
   * - ``ARROW:distinct_count:exact``
     - ``int64``
     - The number of distinct values in the target column. (exact)
   * - ``ARROW:distinct_count:approximate``
     - ``float64``
     - The number of distinct values in the target
       column. (approximate)
   * - ``ARROW:max_byte_width:exact``
     - ``int64``
     - The maximum size in bytes of a row in the target
       column. (exact)
   * - ``ARROW:max_byte_width:approximate``

>      - ``float64``: TODO: Should we use ``int64`` instead?

Author: Related discussion: https://github.com/apache/arrow/pull/43553/files#r1871759234

>      - ``float64``: TODO: Should we use ``int64`` instead?

Author: There is no objection for using `float64`.

So I use `float64`.

       :ref:`ipc-recordbatch-message`.
   * - ``statistics``
     - ``map``
     - ``false``
     - Statistics for the target column, table or record batch. See
       the separate table below for details.

Here is the details of the ``map`` of the ``statistics``:

.. list-table::
   :header-rows: 1

   * - Key or items
     - Data type
     - Nullable
     - Notes
   * - key
     - ``dictionary<values: utf8, indices: int32>``
     - ``false``
     - The string key is the name of the
       statistic. Dictionary-encoding is used for efficiency as the
       same statistic may be repeated for different columns.
       Different keys are assigned for exact and approximate statistic
       values. Each statistic has their own description below.
   * - items
     - ``dense_union``
     - ``false``
     - Statistics value is dense union. It has at least all needed
       types based on statistics kinds in the keys. For example, you
       need at least ``int64`` and ``float64`` types when you have a
       ``int64`` distinct count statistic and a ``float64`` average
       byte width statistic. See the description of each statistic below.

>        Dense union has name for each field but we don't standardize
>        field names for the dense union because we can access to proper
>        field by type code not name. So we can use any valid name for
>        fields.

Reviewer: ```suggestion
       Dense union arrays have names for each field but we don't standardize
       field names for these because we can access the proper
       field by type code instead. So we can use any valid name for
       the fields.
```

Author: Thanks!

Simple record batch
-------------------

Schema::

    vendor_id: int32
    passenger_count: int64

Data::

    vendor_id:       [5, 1, 5, 1, 5]
    passenger_count: [1, 1, 2, 0, null]

Statistics:

.. list-table::
   :header-rows: 1

   * - Target
     - Name
     - Value
   * - Record batch
     - The number of rows
     - ``5``
   * - ``vendor_id``
     - The number of nulls
     - ``0``
   * - ``vendor_id``
     - The number of distinct values
     - ``2``
   * - ``vendor_id``
     - The max value
     - ``5``
   * - ``vendor_id``
     - The min value
     - ``1``
   * - ``passenger_count``
     - The number of nulls
     - ``1``
   * - ``passenger_count``
     - The number of distinct values
     - ``3``
   * - ``passenger_count``
     - The max value
     - ``2``
   * - ``passenger_count``
     - The min value
     - ``0``

Column indexes:

.. list-table::
   :header-rows: 1

   * - Index
     - Target
   * - ``0``
     - ``vendor_id``
   * - ``1``
     - ``passenger_count``

Statistics schema::

    struct<
      column: int32,
      statistics: map<
        key: dictionary<values: utf8, indices: int32>,
        items: dense_union<0: int64>
      >
    >

> Statistics array::

Author: Copied from: https://github.com/apache/arrow/pull/43553/files#r1896750644

I didn't understand this example. I thought the statistics were structs, so I would have expected the data to look something like this (perhaps we could give the "logical contents" and then the specific array encoding):

```
[ 
  // first struct element
  { 
    column: null, # record batch
     statistics: {
        "ARROW:row_count:exact": 0
     }
   },
  { 
    column: 0, # vendor_id
     statistics: {
        "ARROW:null_count:exact": 0,
        "ARROW:distinct_count:exact": 2,
        "ARROW:max_value:exact": 5,
        "ARROW:min_value:exact": 1,
     }
   },
...
]
```

I can help work out the example if people think this is a good idea

Author: Ah, it makes sense.
I used physical representation (columnar representation) here but logical representation (row based representation) may be easy to understand.

How about showing both of them? (We keep the current representation and add row based representation like you suggested.)

3. Module B processes the passed Apache Arrow data.

If module A can pass the statistics associated with the Apache Parquet
file to module B, module B can use the statistics to optimize its
query plan.

For example, DuckDB uses this approach but DuckDB couldn't use
statistics because there wasn't the standardized way to represent
statistics for an Apache Arrow data.

.. seealso::

   `duckdb::ArrowTableFunction::ArrowScanBind() in DuckDB 1.1.3
   <https://github.com/duckdb/duckdb/blob/v1.1.3/src/function/table/arrow.cpp#L373-L403>`_

Goals
-----

* Establish a standard way to represent statistics as an Apache Arrow
  array.

Non-goals
---------

* Establish a standard way to pass an Apache Arrow array that
  represents statistics.
* Establish a standard way to embed statistics to an Apache Arrow
  array.

Schema
======

This specification provides only the schema for statistics. This is
the canonical schema to represent statistics about an Apache Arrow
dataset as Apache Arrow data.

Here is the outline of the schema for statistics::

    struct<
      column: int32,
      statistics: map<
        key: dictionary<
          indices: int32,
          dictionary: utf8

>         >,

Author: OK. Let's use `values` instead of `dictionary`.

FYI: `dictionary` is taken from the specification: https://arrow.apache.org/docs/format/Columnar.html#dictionary-encoded-layout

   * - ``statistics``
     - ``map``
     - ``false``
     - Statistics for the target column, table or record batch. See
       the separate table below for details.

Here is the details of the ``map`` of the ``statistics``:

.. list-table::
   :header-rows: 1

   * - Key or items
     - Data type
     - Nullable
     - Notes
   * - key
     - ``dictionary<indices: int32, dictionary: utf8>``
     - ``false``
     - Statistics key is string. Dictionary is used for
       efficiency. Different keys are assigned for exact value and
       approximate value. Also see the separate description below for
       statistics key.
   * - items
     - ``dense_union``
     - ``false``
     - Statistics value is dense union. It has at least all needed
       types based on statistics kinds in the keys. For example, you
       need at least ``int64`` and ``float64`` types when you have a
       ``int64`` distinct count statistic and a ``float64`` average
       byte width statistic. Also see the separate description below
       for statistics key.

       We don't standardize field names for the dense union because we
       can access to proper field by type code not name. So we can use

>        any valid name for fields.

Author: It makes sense. How about https://github.com/apache/arrow/pull/45058/files#r1896223942 ?

>        any valid name for fields.

Reviewer: Good! I fixed some typos.

       column. (exact)
   * - ``ARROW:max_byte_width:approximate``
     - ``float64``
     - The maximum size in bytes of a row in the target
       column. (approximate)
   * - ``ARROW:max_value:exact``
     - Target dependent
     - The maximum value in the target column. (exact)
   * - ``ARROW:max_value:approximate``
     - Target dependent
     - The maximum value in the target column. (approximate)
   * - ``ARROW:min_value:exact``
     - Target dependent
     - The minimum value in the target column. (exact)
   * - ``ARROW:min_value:approximate``
     - Target dependent
     - The minimum value in the target column. (approximate)
   * - ``ARROW:null_count:exact``
     - ``int64``
     - The number of nulls in the target column. (exact)
   * - ``ARROW:null_count:approximate``
     - ``float64``
     - The number of nulls in the target column. (approximate)
   * - ``ARROW:row_count:exact``
     - ``int64``
     - The number of rows in the target table, record batch or
       array. (exact)
   * - ``ARROW:row_count:approximate``
     - ``float64``
     - The number of rows in the target table, record batch or
       array. (approximate)

> If you find a missing statistics key that is usable for multiple
> systems, please propose it on the `Apache Arrow development
> mailing-list <https://arrow.apache.org/community/>`__.

Reviewer: ```suggestion
If you find a statistic that might be useful to multiple
systems, please propose it on the `Apache Arrow development
mailing-list <https://arrow.apache.org/community/>`__.

Interoperability improves when producers and consumers of
statistics agree on a previously agreed upon statistic
specification.
```

       :ref:`ipc-recordbatch-message`.
   * - ``statistics``
     - ``map``
     - ``false``
     - Statistics for the target column, table or record batch. See
       the separate table below for details.

Here is the details of the ``map`` of the ``statistics``:

.. list-table::
   :header-rows: 1

   * - Key or items
     - Data type
     - Nullable
     - Notes
   * - key
     - ``dictionary<values=utf8, indices=int32>``
     - ``false``
     - The string key is the name of the statistic. Dictionary-encoding is used for
       efficiency as the same statistic may be repeated for different columns.
       Different keys are assigned for exact and
       approximate statistic values. Each statistic has their own description below.
   * - items
     - ``dense_union``
     - ``false``
     - Statistics value is dense union. It has at least all needed
       types based on statistics kinds in the keys. For example, you
       need at least ``int64`` and ``float64`` types when you have a
       ``int64`` distinct count statistic and a ``float64`` average
       byte width statistic. See the description of each statistic below.

>        We don't standardize field names for the dense union because we
>        can access to proper field by type code not name. So we can use
>        any valid name for fields.

Author: ```suggestion
       Dense union has name for each field but we don't standardize field names for the dense union because we
       can access to proper field by type code not name. So we can use
       any valid name for fields.
```
